### DISCUSSION TITLE
Warning logs and strict mode

### DISCUSSION THREAD

**Discussion by pawamoy** [mkdocs/mkdocs#3818] [General]
Since many (if not most) users of MkDocs use strict mode (locally or in CI) to assert there's no issue within their docs, when a plugin updates its logic to emit *more warnings* given the same previous contents, it generally means that builds (can) break for its downstream users. Broken builds are generally seen as a "breaking change", so I'm left wondering whether such changes should warrant bumping the major version number of the plugin.

If we compare with the usual process for introducing breaking changes in public APIs, which is to first release a patch version that emits a deprecation warning, and then after some time a major version with the breaking change, IMO changes in warnings shouldn't warrant a major bump. But both situations are not completely equivalent: what when warnings themselves are the breaking change? :thinking: 

Developers who want to catch deprecated use of their dependencies can opt into making their pytest runs fail when such warnings are emitted. So essentially, if an upstream project releases a patch that emits a new deprecation warning, it makes my pytest runs fail. But that's generally not considered a breaking change, as it only impacts my test runs, and not my software running in production. What about docs? Warnings can make my strict docs builds fail, but technically I can still build and deploy the docs, ignoring the warnings.

So, in the end, should I bump my plugin to a new major version when it starts emitting more warning logs? :thinking: :thinking: :thinking: I want to say no, because otherwise it means I have to somehow create a transition period where I emit an INFO log telling that in the near future warnings will be emitted, release that, then wait a bit, and release another major version (well, the usual dance for public API breaking changes). That's a lot of maintenance for something that actually doesn't prevent users from building and deploying their docs (as long as the warnings are easily actionable by the users) :thinking: 

**EDIT: Poll created here: https://github.com/mkdocs/mkdocs/discussions/3822**

**Comment by louwers** [comment]
I would want my build to fail if there is a legitimate issue with my documentation. That is what strict mode is for after all. I agree with you that it's not worth the effort to consider more warnings a breaking change, even though technically it is.

**Reply by pawamoy** [reply]
Well I'm kinda arguing here that an added warning is technically *not* a breaking change, because just like failing pytest runs, MkDocs' strict mode is opt-in behavior :slightly_smiling_face: Either way, yup, not worth the effort! The exception maybe is when the warnings are hard to fix / to get rid of, in that case it would be nice from the maintainers to offer a transition period :slightly_smiling_face:

**Reply by louwers** [reply]
It is a breaking change because given the same inputs, there is a different output.

That is is opt-in is irrelevant. Most functionality of APIs tends to be opt-in.

**Reply by pawamoy** [reply]
> given the same inputs, there is a different output

Any change would be a breaking change according to this definition (which is true in theory :smile:). Usually we add a bit of nuance to this, by explicitly declaring what our public API is, or what can be considered stable. So the initial question could probably be considered moot, and the answer always be: "it depends on the project", since each project can decide that "yes", this is a breaking change, or "no", this is not a breaking change :relaxed:

**Reply by louwers** [reply]
@pawamoy Not true. A new API is not a breaking change. Nor is changes that do not change the behavior of public APIs, just the implementation.

It is a breaking change, but it probably shouldn't be considered as one.

**Reply by Andre601** [reply]
For me, breaking changes are what isn't backwards compatible with previous iterations... Like when an API method gets removed f.e.
This here seems to not really fall into this definition for me, so I personally do not consider it a breaking change... Tho, it's obviously only how I see it.

**Comment by pawamoy** [comment]
The [poll](https://fosstodon.org/@pawamoy/113039806776516404) on Mastodon says:

> Should new warnings (logs or actual warnings) emitted given the same previous contents (user code/docs didn't change) be considered a breaking change?

- 30% Yes
- 50% No
- 20% Depends

No additional comments to expand on "depends" :sweat_smile: Only 10 voters :man_shrugging:

**Reply by Andre601** [reply]
I didn't know there was even a poll... Primarely because I don't follow your account. Perhapse a Mastodon account for mkdocs itself would be useful to have?
(There was an automated group account, but the site chirp.social was discontinued)

**Reply by pawamoy** [reply]
Yeah I should have posted the link here earlier. I used the `#mkdocs` hashtag, you could follow it. There's a new Mastodon group based on Guppe: https://fosstodon.org/@mkdocs@a.gup.pe. It doesn't feel like posting to the group (3 followers) would have driven more engagement to the poll (or to my other MkDocs-related posts), so I usually just use the hashtag.

**Reply by lovelydinosaur** [reply]
Good reminder. Currently underutilised... https://github.com/mkdocs/mkdocs/discussions/categories/polls

**Reply by pawamoy** [reply]
Right! I'll create a new poll here then, to see if we get more votes :slightly_smiling_face:

**Reply by pawamoy** [reply]
https://github.com/mkdocs/mkdocs/discussions/3822
